### PR TITLE
DEV: Upgrade Rails to 7.1

### PR DESCRIPTION
**Pull request by Flink**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -6,13 +6,15 @@ source "https://rubygems.org"
...
-gem "railties", "< 7.1"
+%w[
+  actionmailer
+  actionpack
+  actionview
+  activemodel
+  activerecord
+  activesupport
+  railties
+].each { |name| gem name, "~> 7.1.0" }
```
Comment (davidtaylorhq) [inline Gemfile:17]: I think this will break dependabot's parsing of the Gemfile 👀

Comment (Flink) [reply Gemfile:17]: You’re probably right, the dependabot code seems to rely on its own logic and not on the bundler API.

File: spec/fixtures/mmdb/mmdb.tar.gz
Comment (davidtaylorhq) [inline spec/fixtures/mmdb/mmdb.tar.gz:1]: 👀 don't think this should be committed

File: spec/support/locales/site_settings.en.yml
Comment (davidtaylorhq) [inline spec/support/locales/site_settings.en.yml:1]: How come this locale stuff is part of the Rails upgrade? Did it somehow get added by accident in a rebase? Or is there a real reason? 🤔

File: spec/fixtures/mmdb/mmdb.tar.gz
Comment (Flink) [reply spec/fixtures/mmdb/mmdb.tar.gz:1]: It’s related to the `spec/lib/discourse_ip_info_spec.rb` spec, it was failing, and I was wondering how it could work without having a proper tar file to decompress. So I just added this file (what’s in it is the other mmdb file we already have in our fixtures), changed the stubbed response to return it and everything’s green. Maybe I’m nissing something else though? :sweat_smile:

File: spec/support/locales/site_settings.en.yml
Comment (Flink) [reply spec/support/locales/site_settings.en.yml:1]: There’s a real reason: https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#config-i18n-raise-on-missing-translations-true-now-raises-on-any-missing-translation.
Previously it would raise only in controllers and views, now it’s everywhere. I changed some related stuff after introducing these locale files though, so let me check if I can remove some of them.

File: spec/fixtures/mmdb/mmdb.tar.gz
Comment (Flink) [reply spec/fixtures/mmdb/mmdb.tar.gz:1]: I’m not sure it has an impact on the code itself but more on the spec. I don’t understand how the current spec (in `main`) can work since there is no file to decompress… I’ll dig into this a little and see if I can find something :sweat_smile:

File: config/initializers/100-logster.rb [hunk 2]
```diff
@@ -10,12 +10,12 @@
 if Rails.env.development? && !Sidekiq.server? && ENV["RAILS_LOGS_STDOUT"] == "1"
   Rails.application.config.after_initialize do
     console = ActiveSupport::Logger.new(STDOUT)
-    original_logger = Rails.logger.chained.first
+    original_logger = Rails.logger.broadcasts.first
```
Comment (tgxworld) [inline config/initializers/100-logster.rb:13]: `chained` is a `Logster::Logger` attribute, are we changing some behaviour here?

File: app/models/theme_setting.rb [hunk 3]
```diff
@@ -54,7 +54,7 @@ def json_value_size
         :json_value,
         I18n.t(
           "theme_settings.errors.json_value.too_large",
-          max_size_megabytes: MAXIMUM_JSON_VALUE_SIZE_BYTES / 1024 / 1024,
+          max_size: MAXIMUM_JSON_VALUE_SIZE_BYTES / 1024 / 1024,
```
Comment (tgxworld) [inline app/models/theme_setting.rb:57]: I think we can pull this fix out into another PR since this is not directly related to the upgrading Rails.

File: config/application.rb [hunk 4]
```diff
@@ -112,8 +112,9 @@ def config.database_configuration
       ENV["DISCOURSE_MULTISITE_CONFIG_PATH"] || GlobalSetting.multisite_config_path
     config.multisite_config_path = File.absolute_path(multisite_config_path, Rails.root)
 
+    config.autoload_lib(ignore: %w[common_passwords emoji generators javascripts tasks])
+    Rails.autoloaders.main.do_not_eager_load(config.root.join("lib"))
```
Comment (tgxworld) [inline config/application.rb:116]: I am curious about the change in Rails 7.1 which requires us to do this now. Will you be able to share more? Thank you 🙇

File: config/initializers/new_framework_defaults_7_1.rb [hunk 5]
```diff
@@ -0,0 +1,224 @@
+# frozen_string_literal: true
```
Comment (tgxworld) [inline config/initializers/new_framework_defaults_7_1.rb:1]: We still have https://github.com/discourse/discourse/blob/main/config/initializers/new_framework_defaults_7_0.rb. Should we work on removing that file first before merging the Rails 7.1 upgrade?

File: lib/discourse.rb [hunk 6]
```diff
@@ -1116,16 +1116,7 @@ def self.preload_rails!
...
-    schema_cache = ActiveRecord::Base.connection.schema_cache
-
     RailsMultisite::ConnectionManagement.safe_each_connection do
-      # load up schema cache for all multisite assuming all dbs have
-      # an identical schema
-      dup_cache = schema_cache.dup
-      # this line is not really needed, but just in case the
-      # underlying implementation changes lets give it a shot
-      dup_cache.connection = nil
-      ActiveRecord::Base.connection.schema_cache = dup_cache
```
Comment (tgxworld) [inline lib/discourse.rb:1128]: I am curious about the reason for making this change. Is it because the Rails API is no longer supported? How does it affect preloading if this change is removed?

File: config/initializers/100-logster.rb (hunk 2 above)
Comment (Flink) [reply config/initializers/100-logster.rb:13]: Yes, because `Rails.logger` is not a `Logster::Logger` anymore with the changes introduced in Rails 7.1. The Logster logger is added to the broadcast logger, like any other logger.
You can see the change I made some time ago to Logster for Rails 7.1 compatibility: https://github.com/discourse/logster/commit/a023d778fb28bfa731865a49ca4383b1b50a9117

File: config/application.rb (hunk 4 above)
Comment (Flink) [reply config/application.rb:116]: It’s what’s recommended here when upgrading: https://guides.rubyonrails.org/upgrading_ruby_on_rails.html#config-autoload-lib-and-config-autoload-lib-once. They say it should still work, but it’s probably better to follow the new convention :slightly_smiling_face:

File: config/initializers/new_framework_defaults_7_1.rb (hunk 5 above)
Comment (Flink) [reply config/initializers/new_framework_defaults_7_1.rb:1]: I can try and see how it goes, yes.

File: lib/discourse.rb (hunk 6 above)
Comment (Flink) [reply lib/discourse.rb:1128]: Indeed, you can’t set the schema cache on the connection anymore. There’s a place where it might be feasible, but it’s buried relatively deep (on the pool config object) and that method has been removed from the current Rails code. From what I understand, though, the schema cache is automatically created per connection once. So I guess the worst-case scenario is that it will be fetched when the first connection is established (for each connection) instead of sharing the same one here like we were doing. I would tend to say it’s probably not that impactful, but I can’t really give you a definitive answer here.

File: app/models/theme_setting.rb (hunk 3 above)
Comment (Flink) [reply app/models/theme_setting.rb:57]: Done in https://github.com/discourse/discourse/pull/27450

File: config/initializers/new_framework_defaults_7_1.rb (hunk 5 above)
Comment (Flink) [reply config/initializers/new_framework_defaults_7_1.rb:1]: Done! :grin:

File: lib/hijack.rb [hunk 7]
```diff
@@ -53,10 +53,10 @@ def hijack(info: nil, &blk)
           # this trick avoids double render, also avoids any litter that the controller hooks
           # place on the response
           instance = controller_class.new
-          response = ActionDispatch::Response.new
-          instance.response = response
+          response = ActionDispatch::Response.new.tap { _1.request = request_copy }
+          instance.set_response!(response)
+          instance.set_request!(request_copy)
```
Comment (CvX) [inline lib/hijack.rb:58]: Aren't `_1.request = request_copy` and `instance.set_request!(request_copy)` doing the same thing?

**Review by CvX — APPROVED** [review]

File: lib/hijack.rb (hunk 7 above)
Comment (Flink) [reply lib/hijack.rb:58]: That’s what I was thinking at first, and `instance.set_request!(request_copy)` wasn’t there initially. But some of the code was failing. I don’t remember exactly why, but IIRC, sometimes things are delegated to the response object, and sometimes it’s directly in the request object set on the controller.
